### DISCUSSION TITLE
Changing the one line bash script install to support multiple platforms

### DISCUSSION THREAD

**Discussion by pottekkat** [meshery/meshery#3168] [Ideas]
Recently the [bash script](https://github.com/layer5io/meshery/blob/master/meshery) to install mesheryctl and start Meshery was updated to allow the users specify a platform to deploy Meshery to (Docker and Kubernetes).

Users are able to pass in the flags as shown below(this is a work around)-

![Screenshot from 2021-06-08 23-31-39](https://user-images.githubusercontent.com/49474499/121235287-51615800-c8b2-11eb-848d-5fbd27504cb7.png)

But, if the user does not specify a platform, it will throw an error(which it should not)-

![Screenshot from 2021-06-08 23-37-19](https://user-images.githubusercontent.com/49474499/121235405-7b1a7f00-c8b2-11eb-8b42-cc5834436726.png)

These errors and workarounds are there because with the current structure of the one-line installer- `curl -L https://git.io/meshery | bash -`, we would not be able to pass in flags or get an input from the user.

The error "Invalid platform: " is because, when the user does not specify a platform, the bash script reads from the standard input. But, this is already taken up by the script hence causing the platform to be " " empty.

This can be fixed by changing the one liner as shown below-

![Screenshot from 2021-06-05 11-41-30](https://user-images.githubusercontent.com/49474499/121239590-0d248680-c8b7-11eb-8df3-25817f0a963b.png)

We can use the `-p` flag to pass in the platform.

And, if the user does not specify a platform, it will prompt the user to enter a platform and validates it-

![Screenshot from 2021-06-05 11-40-54](https://user-images.githubusercontent.com/49474499/121239676-2c231880-c8b7-11eb-9946-35b5655f8943.png)

This discussion is to decide whether we should switch to the new one liner that I'm proposing-

For Docker-
```
bash -c "$(curl -L https://git.io/meshery)" -- -p docker
```

For Kubernetes-
```
bash -c "$(curl -L https://git.io/meshery)" -- -p kubernetes
```

Without specifying platform and prompting the user-
```
bash -c "$(curl -L https://git.io/meshery)"
```

Please share what you think and how we should proceed.

**Comment by leecalcote** [comment]
Alternatively, we update all references to the deprecated one-liner, so that the PLATFORM var is "always" included and when it is not included (for whatever reason) that the user is prompted for their desired platform. To assist in avoiding broken programmatic invocations of this script, a helpful return message could point out that no "PLATFORM" preference was provided.

Changing to from `curl` piped to `bash` seems like a bigger change than is necessary. Is your primary motivation here to avoid having to update existing script invocations; to avoid breaking those?

**Reply by pottekkat** [reply]
The reason for this change is that when we pipe the bash script-

> The error "Invalid platform: " is because, when the user does not specify a platform, the bash script reads from the standard input. But, this is already taken up by the script hence causing the platform to be " " empty.

- we cannot get an input from the user. Making the switch as I suggested will make it easy to specify the platform with the `-p` flag and also make it interactive(get platform from user and validate it if they don't specify it).

// @leecalcote

**Reply by dhruv0000** [reply]
Can't we have different URL to predefine the platform.

```bash -c "$(curl -L https://git.io/meshery-k8s)"```
```bash -c "$(curl -L https://git.io/meshery-docker)"```

And ```bash -c "$(curl -L https://git.io/meshery)"``` to give a prompt?

**Reply by leecalcote** [reply]
@navendu-pottekkat we spoke about this previously and noted that there should always be a PLATFORM provided in the instructions (the install command) given to users. If the user neglects to include PLATFORM and has a malfunctioning install, while that is their fault, there are things that we can put into place to help the install overcome missing information: 1) prompt them, 2) provide reasonable defaults (Docker now and eventually Kubernetes as the default platform) 3) provide them with helpful output, informing them that their installation command is malformed. 

We could provide a timeout on the interactive prompt and move forward with the default platform if the user doesn't answer in 30 seconds. So, in essence, we would give them an automatic silent install with initially, optional interactive choice.

This proposal does not overcomes the crux of the issue described any differently than the current implementation provides. This proposal simply moves the issue from a lack of PLATFORM to a lack of -p.

**Reply by pottekkat** [reply]
If the user neglects to include PLATFORM, we indeed prompt the user. But this prompt will always fail because of [this](https://github.com/layer5io/meshery/discussions/3168#discussioncomment-845615).

Inorder to handle missing platform, we have to move away from piping the script to bash. **We cannot use interactive prompt and piping the script because both use the same standard input to read.**

Hence, what this change will give is the **ability to have interactive prompts.**

**Reply by pottekkat** [reply]
@dhruv0000 that would mean having multiple bash scripts right? One for each platform?

**Reply by hexxdump** [reply]
> @dhruv0000 that would mean having multiple bash scripts right? One for each platform?

yes, does not make sense to have multiple scripts for multiple platforms, where most of the instructions are same.

**Reply by hexxdump** [reply]
> If the user neglects to include PLATFORM, we indeed prompt the user. But this prompt will always fail because of [this](https://github.com/layer5io/meshery/discussions/3168#discussioncomment-845615).
> 
> Inorder to handle missing platform, we have to move away from piping the script to bash. **We cannot use interactive prompt and piping the script because both use the same standard input to read.**
> 
> Hence, what this change will give is the **ability to have interactive prompts.**

@navendu-pottekkat , if we plan to move away from pipe, we could achieve the user prompt with the two commands syntax I mentioned earlier. 
`curl -L https://git.io/meshery -o meshery;  PLATFORM=docker ./meshery` -- for Docker platform
`curl -L https://git.io/meshery -o meshery;  PLATFORM=kubernetes ./meshery` -- for kubernetes platform
`curl -L https://git.io/meshery -o meshery; ./meshery`  -- prompts for user input

**Reply by devansh42** [reply]
I think, we can't use this syntax, as this requires **meshery** to be an executable file.

**Reply by devansh42** [reply]
Actually, Piping is quite difficult to understand, it just complicates a simple command, and as Lee said that PLATFORM and --p are equivalent in some nature and both would have some consequences if omitted by user. But non-pipe version can help us to extend the functionality of the installer in the future. Suppose in the future we add some new options then for the piped version it'd require to add new variables for each of them that will increase the cognitive complexity of the command, where non-piped version looks pretty mainstream, we can add many options with small argument name.
e.g. **-p** is easy to use and remember than **PLATFORM** or **-n** is simpler than **DEPLOY_MESHERY**.
